### ISSUE TITLE
manually specify powershell version (e.g. 7)

### ISSUE THREAD

**Issue by OneCyrus** [Azure/azure-functions-powershell-worker#371]
is there a way to change the used powershell version? would like to give powershell 7 rc1 a try as we have some code which doesn't work with PS6.

**Comment by JustinGrote** [comment]
The azure functions core tools (the one you run locally) might be able to do it but you'd have to modify the source code, there's no built in option.

No roadmap's been published yet for upgrading the worker to 7, I imagine that will coincide with the upgrade from functions runtime v2 to v3.

**Comment by AnatoliB** [comment]
We will allow switching to PS7 soon after PS7 goes GA.

Right now, the only way to use PS7 in Functions is to package/deploy the PS7 files together with your app and invoke pwsh.exe from your Function code (which I would recommend only if you have a very strong reason).

By the way, Functions runtime v3 is currently released, and it also uses PS6.

**Comment by dpolivy** [comment]
@AnatoliB Now that PS7 is GA, any timeline for being able to use this with Azure Functions v3 runtime? I see the code appears to be committed to the repo, but not sure how to use it either locally or in Azure Functions without building myself?

**Comment by AnatoliB** [comment]
@dpolivy Right now, building the worker from the dev branch is the only way. Expect PS7 to be available on Azure between the end of March and the middle of April. I will be updating this thread with more precise ETA and status as soon as I have them.

**Comment by JustinGrote** [comment]
@dpolivy you can also download the built nupkg from Appveyor, rename it to .zip, and then extract it from there
https://ci.appveyor.com/project/appsvc/azure-functions-powershell-worker/builds/31432379/job/0bend5twai5dkg3e/artifacts
![image](https://user-images.githubusercontent.com/15258962/76661437-ad66f080-6538-11ea-9c70-e7b8fe4257eb.png)

Extract it to your workers directory for core tools per the readme. I installed azure-functions-core-tools using http://scoop.sh so mine is at 
C:\Users\jgrote\scoop\apps\azure-functions-core-tools\current\workers

**Comment by AnatoliB** [comment]
@dpolivy @JustinGrote Thank you, Justin. If you do this, you also need to set the following variable before starting the function host locally:
``` PowerShell
$env:FUNCTIONS_WORKER_RUNTIME_VERSION = '~7'
```
You will not have to do this when we release PS7 support officially.

**Comment by JustinGrote** [comment]
@anatolib we are now at mid-april, can you provide a revised ETA? :)

**Comment by AnatoliB** [comment]
> @AnatoliB we are now at mid-april, can you provide a revised ETA? :)

The situation is currently complicated by certain events beyond our control. The coding work is finished, but the deployment to Azure will probably happen around _the middle of May_. I'm sorry I don't have a more precise and reliable ETA, but this is the best I can do at this point. Thank you for your patience!

**Comment by JustinGrote** [comment]
@anatolib Understood! Thank you for all your hard work and I appreciate the update, it just helps to know how far out I should be looking at refactoring.

**Comment by akhilthomas011** [comment]
@AnatoliB Is this in production now? Can you help me with the status?

**Comment by AnatoliB** [comment]
The Functions runtime with PowerShell 7 is being rolled out to production now, I expect it to complete by June 5. I will update this thread when complete.

**Comment by akhilthomas011** [comment]
@AnatoliB, Really appreciate it! Thanks!

**Comment by JustinGrote** [comment]
@AnatoliB FYI I saw this today I've never seen before, may be related to rollout:
https://twitter.com/JustinWGrote/status/1266033654016311296

**Comment by wsf11** [comment]
Hi @AnatoliB , any update on when this will be in production? Thanks for working on this!

**Comment by AnatoliB** [comment]
The runtime rollout is almost finished and PowerShell 7 is available in many regions now, expected to finish the rest tomorrow (6/9).

Portal and tools support will take a bit longer. In the meantime, in order to upgrade your app to PowerShell 7, set the `powerShellVersion` property on the `/subscriptions/{subId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{appName}/config/web` resource to `~7`. You can do it with the following command:

``` PowerShell
$SubId = ...
$ResourceGroupName = ...
$AppName = ...
Set-AzResource -ResourceId "/subscriptions/$SubId/resourceGroups/$ResourceGroupName/providers/Microsoft.Web/sites/$AppName/config/web" -UsePatchSemantics -Properties @{ powerShellVersion = '~7' } -Force
```

Any other way of updating ARM resources (Resource Explorer, ARM templates, etc.) should also work. Set the property to `~6` to revert the app back to PowerShell 6, if required.

You may need to restart the app after this.

~~A **temporary caveat**: this property is currently _write-only_. You can set it, but GET will return null, regardless of the actually persisted value. This bug will be fixed soon. This problem does not affect the language version used for executing functions.~~

**Comment by JustinGrote** [comment]
@AnatoliB great news, can you have them make a post on the Functions Blog about it when it's fully rolled out?

**Comment by AnatoliB** [comment]
@JustinGrote We will announce this officially only after the Portal and tools support catch up. Until then, I will just leave these instructions here.

**Comment by AnatoliB** [comment]
The rollout is complete now, PowerShell 7 is available in all the public regions. I'm closing this issue, but I will be updating it with the tools and portal status. In the meantime, please use the instructions above, and we will appreciate your feedback.

**Comment by AnatoliB** [comment]
The rough ETA for official GA (including portal and tools support) is the beginning of July.

**Comment by nemanja-jovic** [comment]
Hi! Very appreciated for taking the time to roll this out. At the moment I have tried to set the powershellVersion via Powershell cmdlet that @AnatoliB posted, I have also tried deploying the app service setting via ARM. Still I get the following error when trying to import one of my custom modules which require version 7.
`[Error] EXCEPTION: Import-Module : The version of PowerShell on this computer is '6.2.4'.`
Any thoughts?
